### ISSUE TITLE
torch_geometric.nn.aggr on Google Colab

### ISSUE THREAD

**Issue by RafiBrent** [pyg-team/pytorch_geometric#5126]
### 😵 Describe the installation problem

I have successfully installed the following:

!pip install torch-scatter -f https://data.pyg.org/whl/torch-1.12.0+cu113.html
!pip install torch-sparse -f https://data.pyg.org/whl/torch-1.12.0+cu113.html
!pip install torch-geometric -f https://data.pyg.org/whl/torch-1.12.0+cu113.html

and tried various modifications and reformulations of the environment as well. In all of these, the majority of functionality was present, but the notebook did not recognize the torch_geometric.nn.aggr module. I am positive that my import statement was phrased correctly, as it was copied directly from examples in the source code. Any advice on how to address this error would be greatly appreciated.

### Environment

* PyG version: 2.0.3
* PyTorch version: 1.12.0
* OS: Google Colab
* Python version: 3.7.13
* CUDA/cuDNN version: 11.1
* How you installed PyTorch and PyG (`conda`, `pip`, source): See above
* Any other relevant information (*e.g.*, version of `torch-scatter`): torch-scatter is 2.0.9, torch-sparse is 0.6.14

**Comment by rusty1s** [comment]
The `nn.aggr` package just got added (we plan to release it next week). Try [installing PyG nightly or from master](https://github.com/pyg-team/pytorch_geometric#nightly-and-master).

**Comment by RafiBrent** [comment]
Thanks, I appreciate your help. This is a super awesome package-- thanks so much for adding it!

**Comment by RafiBrent** [comment]
Hey, sorry to reopen this thread-- I just wanted to double-check whether these modules are set up to deal with batched data. I'm getting issues in that context with EquilibriumAggregation and it looks like there's nothing batch-related in the source code. If this is not currently a feature of Aggregation modules, it would be super helpful to have that included in some future version. Thanks again for your responsiveness to comments and for all of your work on PyG!

**Comment by rusty1s** [comment]
Yes, all aggregations can also be applied in batch-mode on varying graph sizes. The `index` tensor ensures that no data is shared across examples.

**Comment by RafiBrent** [comment]
Ok, thanks. Should this index tensor be something such as data.batch, indicating which node belongs in which graph? I've tried that but am getting the same output (all zeroes) for every run of EquilibriumAggregation. I copy-pasted the source code to help myself debug what's going wrong, and it looks like the output of the ResNet is changing (as it should) but the value of grad (in MomentumOptimizer) is always a matrix of all zeros. I apologize for my lack of familiarity with PyG, but do you have any suggestions for what I might be doing wrong? Thanks so much for any advice you may have.

**Comment by rusty1s** [comment]
Yes, `index` refers to a more general concept if aggregations that can be re-used across node-level and graph-level aggregations. For graph-level classification, `index==batch`.

Regarding your problems with `EquilibriumAggregation`, can you share some reproducible example? @Padarn for additional thoughts.

**Comment by Padarn** [comment]
Hey, @RafiBrent I'm happy to try and help with `EquilibriumAggregation`. It could be that there is a bug in the batched version, so if you have some examples that reproduce your problem it would help a lot in figuring it out.

**Comment by gregory-kyro** [comment]
Hey @rusty1s and @Padarn, thank you for helping! I am working with @RafiBrent, so I can reply in his place to provide you more information. 

We are assigning `EquilibriumAggregation`  to our SG-CNN model according to: ` self.eq_aggr = EquilibriumAggregation(12, 12, [200,100,12], grad_iter=5)` where `12, 12` is (size_of_propagation_output, size_of_propagation_output).

Within the forward function of our propagation step, we aggregate the output of the propagation with : `pool_x = self.eq_aggr(x=non_covalent_ligand_only_x.to(device), index = data.batch)` where `non_covalent_ligand_only_x` is an array of size (number_of_nodes_in_batch * size_of_propagation_output), `data` is a class that contains (x=[number_of_nodes_in_batch, number_of_node_features], edge_index, edge_attribute, y=label, batch=batch, ptr=ptr). `data.batch` is a tensor of size (number_of_nodes_in_batch), where the elements are the assigned batch index (in our case, either 0,1,2,3,4,5,6, or 7). 

The source code of class `MomentumOptimizer`:

```
class MomentumOptimizer(torch.nn.Module):

    Provides an inner loop optimizer for the implicitly defined output
    layer. It is based on an unrolled Nesterov momentum algorithm.

    Args:
        learning_rate (flaot): learning rate for optimizer.
        momentum (float): momentum for optimizer.
        learnable (bool): If :obj:`True` then the :obj:`learning_rate` and
            :obj:`momentum` will be learnable parameters. If False they
            are fixed. (default: :obj:`True`)

    def __init__(self, learning_rate: float = 0.1, momentum: float = 0.9,
                 learnable: bool = True):
        super().__init__()

        self._initial_lr = learning_rate
        self._initial_mom = momentum
        self._lr = torch.nn.Parameter(Tensor([learning_rate]),
                                      requires_grad=learnable)
        self._mom = torch.nn.Parameter(Tensor([momentum]),
                                       requires_grad=learnable)
        self.softplus = torch.nn.Softplus()
        self.sigmoid = torch.nn.Sigmoid()

    def reset_parameters(self):
        self._lr.data.fill_(self._initial_lr)
        self._mom.data.fill_(self._initial_mom)

    @property
    def learning_rate(self):
        return self.softplus(self._lr)

    @property
    def momentum(self):
        return self.sigmoid(self._mom)

    def forward(
        self,
        x: Tensor,
        y: Tensor,
        index: Optional[Tensor],
        func: Callable[[Tensor, Tensor, Optional[Tensor]], Tensor],
        iterations: int = 5,
    ) -> Tuple[Tensor, float]:
        momentum_buffer = torch.zeros_like(y).to(device)
        for _ in range(iterations):
            val = func(x, y, index)
            grad = torch.autograd.grad(val, y, create_graph=True,
                                       retain_graph=True)[0]
            delta = self.learning_rate * grad.to(device)
            momentum_buffer = self.momentum * momentum_buffer - delta
            y = y.to(device) + momentum_buffer
        return y
```

If we print `grad` during model training, we get a 2D tensor of size (number_of_batches * size_of_propagation_output), and all of the elements are 0. The values do not update throughout the training and remain at 0. Please let me know if you have any idea what is going wrong here. Thank you for all of your help.

**Comment by Padarn** [comment]
Thanks for the details. It has helped me find a silly bug that was introduced (by me 😞 ): I've fixed it in https://github.com/pyg-team/pytorch_geometric/pull/5155, and tested the example to make sure it is now converging.

Its a very minor change, perhaps you can change it in your local install until we get it merged.

**Comment by rusty1s** [comment]
Thanks for fixing @Padarn. Merged it.

**Comment by gregory-kyro** [comment]
@Padarn @rusty1s Thank you for all of the help! I made the change in the source code I have installed locally and it fixed the problem. I really appreciate it :smiley:
